src/util/Coroutine.hpp

```diff
@@ -0,0 +1,207 @@
+
+    /**
+     * @brief Private constructor to create a Coroutine instance.
+     * @param yield The Boost.Asio yield_context for this coroutine.
+     * @param signal A shared signal used for propagating cancellation requests among related coroutines.
+     * @param generation The generation number of this coroutine, used to manage parent/child relationships.
+     */
+    explicit Coroutine(
+        boost::asio::yield_context&& yield,
+        std::shared_ptr<FamilyCancellationSignal> signal = std::make_shared<FamilyCancellationSignal>(),
+        size_t generation = 0
+    );
+
+public:
+    /**
+     * @brief Destructor for the Coroutine.
+     * Handles cleanup, such as disconnecting from the cancellation signal.
+     */
+    ~Coroutine();
+
+    Coroutine(Coroutine const&) = delete;
+    Coroutine(Coroutine&&) = delete;
+
+    Coroutine&
+    operator==(Coroutine&&) = delete;
+
+    Coroutine&
+    operator==(Coroutine const&) = delete;
+
+    /**
+     * @brief Spawns a new top-level coroutine.
+     * @tparam ExecutionContext The type of the I/O execution context (e.g., boost::asio::io_context).
+     * @tparam Fn The type of the invocable function that represents the coroutine body.
+     * @param ioContext The I/O execution context on which to spawn the coroutine.
+     * @param fn The function to be executed as the coroutine. It will receive a Coroutine& argument.
+     */
+    template <typename ExecutionContext, CoroutineFunction Fn>
+    static void
+    spawnNew(ExecutionContext& ioContext, Fn&& fn)
+    {
+        boost::asio::spawn(ioContext, [fn = std::forward<Fn>(fn)](boost::asio::yield_context yield) {
+            Coroutine thisCoroutine{std::move(yield)};
+            fn(thisCoroutine);
+        });
+    }
+
+    /**
+     * @brief Spawns a child coroutine from this coroutine.
+     * The child coroutine shares the same cancellation signal and has an incremented generation number.
+     * @tparam Fn The type of the invocable function that represents the child coroutine body.
+     * @param fn The function to be executed as the child coroutine. It will receive a Coroutine& argument.
+     */
+    template <CoroutineFunction Fn>
+    void
+    spawnChild(Fn&& fn)
+    {
+        if (isCancelled_)
+            return;
+
+        boost::asio::spawn(
+            yield_,
+            [nextGeneration = generation_ + 1,
+             signal = familySignal_,
+             fn = std::forward<Fn>(fn)](boost::asio::yield_context yield) mutable {
+                Coroutine coroutine(std::move(yield), std::move(signal), nextGeneration);
+                fn(coroutine);
+            }
+        );
+    }
+
+    /**
+     * @brief Returns the error code, if any, associated with the last operation in this coroutine.
+     * @return A boost::system::error_code indicating the status.
+     */
+    boost::system::error_code
+    error() const;
+
+    /**
+     * @brief Cancels this specific coroutine and its direct children.
```

Comment:
`Cancels this specific coroutine` this seems wrong

Reply:
I also thought it is wrong, but looking at Coroutine's constructor I can say that the documentation is correct. Probably I should rename this method to `cancel()`.